mlir/lib/Tools/PDLL/Parser/Parser.cpp

```diff
@@ -956,10 +980,24 @@ LogicalResult Parser::parseInclude(SmallVectorImpl<ast::Decl *> &decls) {
   // Check the type of include. If ending with `.pdll`, this is another pdl file
   // to be parsed along with the current module.
   if (filename.ends_with(".pdll")) {
-    if (failed(lexer.pushInclude(filename, fileLoc)))
+    std::string includedFile;
+    if (!lexer.getSourceMgr().OpenIncludeFile(filename.str(), includedFile))
       return emitError(fileLoc,
                        "unable to open include file `" + filename + "`");
 
+    // Check if included has been already processed
+    if (alreadyIncludedFiles.count(includedFile)) {
+      if (lexer.isLexingMainFile()) {
+        emitWarning(
+            fileLoc,
+            "included file is already included, and therefore has no effect");
+      }
+      return success();
+    }
+
+    auto couldPush = lexer.pushInclude(filename, fileLoc);
+    assert(succeeded(couldPush) && "Already validated on OpenIncludeFile");
```

Comment:
I'd prefer to still `return emitError()` here, because pushInclude might have more error cases in the future

Reply:
Done